source/pr_cmds.c

```diff
@@ -1487,330 +1487,286 @@ This is where the magic happens
-int opensetRef[MAX_WAYPOINTS];//Reference values of open list
-int opensetLength;//equivalent of javaScript's array[].length;
+
 #define MaxZombies 16
 
+
+#define WAYPOINT_SET_NONE 	0
+#define WAYPOINT_SET_OPEN 	1
+#define WAYPOINT_SET_CLOSED	2
+
+
+char waypoint_set[MAX_WAYPOINTS]; // waypoint_set[i] contains the set identifier for the i-th waypoint
+unsigned short openset_waypoints[MAX_WAYPOINTS]; // List of waypoints currently in the open set sorted by heuristic cost (index 0 contains lowest cost waypoint)
+unsigned short openset_length; // Current length of the open set
 zombie_ai zombie_list[MaxZombies];
-//Debug//
-void printSortedOpenSet()
-{
-	Con_Printf("Sorted!: ");
-	int qr;
-	for(qr = 0; qr < opensetLength; qr++)
-	{
-		Con_Printf("%i, ",(int)waypoints[openset[qr]].f_score);
+
+
+//
+// Debugs prints the current sorted list of waypoints in the open set
+//
+void sv_way_print_sorted_open_set() {
+	Con_Printf("Sorted open-set F-scores: ");
+	for(int i = 0; i < openset_length; i++) {
+		Con_Printf("%.0f, ",waypoints[openset_waypoints[i]].f_score);
 	}
 	Con_Printf("\n");
 }
-//------//
 
 
-void RemoveWayFromList (int listnumber, int waynum)
-{
-	if(listnumber == 1)
-	{
-		//Con_DPrintf ("RemoveWayFromList: closedset[%i] = %i\n", waynum, 0);
-		closedset[waynum] = 0;
+
+// 
+// Removes a waypoint from a set, if it belongs to it. 
+//
+void sv_way_remove_way_from_set(char set, int waypoint_idx) {
+	// If the waypoint doesn't belong to the current set, stop
+	if(waypoint_set[waypoint_idx] != set) {
 		return;
 	}
-
-	int i;
-	int s;
-	if(listnumber == 2)
-	{
-		for(i = 0; i < opensetLength; i++)
-		{
-			if(openset[i] == waynum)
-			{
-				openset[i] = 0;
-				opensetRef[waynum] = 0;
-
-				for(s = i; s < opensetLength; s++)
-				{
-					openset[s] = openset[s+1];
+	// If removing from open set, also remove from open-set sorted list
+	if(set == WAYPOINT_SET_OPEN) {
+		for(int i = 0; i < openset_length; i++) {
+			if(openset_waypoints[i] == waypoint_idx) {
+				// Shift down all openset entries above this index
+				for(int j = i; j < openset_length - 1; j++) {
+					openset_waypoints[j] = openset_waypoints[j+1];
 				}
-				opensetLength -= 1;
-				return;
+				openset_length -= 1;
+				break;
 			}
 		}
 	}
+	waypoint_set[waypoint_idx] = WAYPOINT_SET_NONE;
 }
 
-void CompareOpenLists()
-{
-	int refCount, count;
-	refCount = 0;
-	count = 0;
-	int i;
-	for(i = 0; i < MAX_WAYPOINTS; i++)
-	{
-		if(openset[i])
-			count++;
-		if(opensetRef[i])
-			refCount++;
+
+//
+// Debug method to verify that `openset` and `opensetRef` remain synchronized
+//
+void sv_way_compare_open_set_lists() {
+	// Count the number of waypoints in the open set
+	int n_openset_waypoints = 0;
+	for(int i = 0; i < n_waypoints; i++) {
+		if(waypoint_set[i] == WAYPOINT_SET_OPEN) {
+			n_openset_waypoints += 1;
+		}
+	}
+
+	if(n_openset_waypoints != openset_length) {
+		Con_Printf("%i%i%i\n", n_openset_waypoints, openset_length);
```

Comment:
`Con_Dprintf`

Reply:
Ditto as above (This is in another debug function that's not being called)